hermes/probe/monitor.go

```diff
@@ -12,9 +12,160 @@
+func (mp *MonitorProbe) Init(name string, opts *options.Options) error {
+	// Cast from Cloudprober defined interface to HermesProbeDef.
+	// This allows for accessing the variables and methods of a HermesProbeDef object.
+	conf, ok := opts.ProbeConf.(*probepb.HermesProbeDef)
+	if !ok {
+		return fmt.Errorf("invalid argument: opts.ProbeConf is not of type *probepb.HermesProbeDef")
+	}
+	mp.name = name
+	mp.config = conf
+
+	mp.targets = make([]*Target, 0)
+	for _, t := range mp.config.GetTargets() {
+		run := &Target{
+			Target: t,
+			Journal: &journalpb.StateJournal{
+				Intent:    &journalpb.Intent{},
+				Filenames: make(map[int32]string),
+			},
+		}
+		mp.targets = append(mp.targets, run)
+	}
+	mp.opts = opts
+	mp.logger = opts.Logger
+
+	return nil
+}
+
+// Start runs the probe indefinitely, unless cancelled, at the configured interval.
+// Probe metrics will be sent via the metricChan at the end of the probe run.
+// This is a required method to implement the cloudprober.Probes.Probe interface.
+// Arguments:
+//	- ctx: context provided for cancelling probe.
+//	- metricChan: unidirectional channel used for sending metrics to be surfaced.
+func (mp *MonitorProbe) Start(ctx context.Context, metricChan chan<- *cpmetrics.EventMetrics) {
+	probeTicker := time.NewTicker(mp.opts.Interval)
+
+	for {
+		select {
+		case <-ctx.Done():
+			probeTicker.Stop()
+			return
+		case <-probeTicker.C:
+			mp.runProbe(ctx, metricChan)
+		}
+	}
+}
+
+// reportMetrics sends the metrics recorded in the current probe run to Cloudprober.
+// Arguments:
+//	- run: metrics from a probe run on a target.
+//	- metricChan: metric channel passed from Cloudprober.
+func reportMetrics(run *metrics.Metrics, metricChan chan<- *cpmetrics.EventMetrics) {
+	for _, op := range run.ProbeOpLatency {
+		for _, m := range op {
+			metricChan <- m
+		}
+	}
+
+	for _, call := range run.ApiCallLatency {
+		for _, m := range call {
+			metricChan <- m
+		}
+	}
+}
+
+// runProbe runs probe for all targets and update EventMetrics.
+// Arguments:
+//	- ctx: pass context to allow for complete cancellation of the probe.
+//	- metricChan: pass the metrics channel for surfacing metrics to Cloudprober.
+func (mp *MonitorProbe) runProbe(ctx context.Context, metricChan chan<- *cpmetrics.EventMetrics) {
+	var wg sync.WaitGroup
+	for _, t := range mp.targets {
+		wg.Add(1)
+		go func(t *Target) {
+			defer wg.Done()
+			var err error
+			if t.LatencyMetrics, err = metrics.NewMetrics(mp.config, t.Target); err != nil {
+				mp.logger.Errorf(err.Error())
+				return
+			}
+			probeCtx, _ := context.WithDeadline(ctx, time.Now().Add(mp.opts.Timeout)) //TODO: Change to mp.Config. ...
+			start := time.Now()
+			exitStatus, err := mp.runProbeForTarget(probeCtx, t)
+			if err != nil {
+				mp.logger.Errorf(err.Error())
+				t.LatencyMetrics.ProbeOpLatency[metrics.ProbeOps[metrics.TOTAL_PROBE_RUN]][metrics.ExitStatuses[exitStatus]].Metric("latency").AddFloat64(time.Now().Sub(start).Seconds())
```

Comment:
Again, using the numeric ID in the maps would remove the need for this translation.

Reply:
Ack.